op-chain-ops/genesis/layer_two_test.go

```diff
@@ -94,3 +93,107 @@ func TestBuildL2DeveloperGenesisDevAccountsFunding(t *testing.T) {
 	require.NoError(t, err)
 	require.Equal(t, 2321, len(gen.Alloc))
 }
+
+func TestBuildL2MainnetGenesis(t *testing.T) {
+	config, err := genesis.NewDeployConfig("./testdata/test-deploy-config-devnet-l1.json")
+	require.Nil(t, err)
+	require.Equal(t, true, config.EnableGovernance, "governance should be enabled")
+
+	backend := backends.NewSimulatedBackend(
+		core.GenesisAlloc{
+			crypto.PubkeyToAddress(testKey.PublicKey): {Balance: big.NewInt(10000000000000000)},
+		},
+		15000000,
+	)
+	block, err := backend.BlockByNumber(context.Background(), common.Big0)
+	require.NoError(t, err)
+
+	gen, err := genesis.BuildL2MainnetGenesis(config, block)
+	require.Nil(t, err)
+	require.NotNil(t, gen)
+
+	proxyBytecode, err := bindings.GetDeployedBytecode("Proxy")
+	require.NoError(t, err)
+
+	for name, predeploy := range predeploys.Predeploys {
+		addr := *predeploy
+
+		account, ok := gen.Alloc[addr]
+		if predeploys.IsDeprecated(addr) && !predeploys.IsProxied(addr) {
+			require.Equal(t, false, ok, name)
+			continue
+		}
+		require.Equal(t, true, ok, name)
+		require.Greater(t, len(account.Code), 0)
+
+		if !predeploys.IsProxied(addr) {
+			require.NotEqual(t, proxyBytecode, account.Code)
+			continue
+		}
+
+		adminSlot, ok := account.Storage[genesis.AdminSlot]
+		require.Equal(t, true, ok)
+		require.Equal(t, predeploys.ProxyAdminAddr.Hash(), adminSlot)
+		require.Equal(t, proxyBytecode, account.Code)
+	}
+	require.Equal(t, 2063, len(gen.Alloc)) // TODO: confirm 2063 is correct!
+
+	if writeFile {
+		file, _ := json.MarshalIndent(gen, "", " ")
+		_ = os.WriteFile("genesis.json", file, 0644)
+	}
+}
+
+// Same test as TestBuildL2MainnetGenesis, only we blow away the governance token config and
+// confirm the governance predeploy doesn't exist (or more precisely, there's an unused proxy
+// contract at its address instead).
+func TestBuildL2MainnetNoGovernanceGenesis(t *testing.T) {
```

Comment:
Perhaps could combine some of the business logic of the 2 tests added

Reply:
done.